File: src/main/java/io/github/hdzitao/editstarters/ui/swing/TableValue.java
package io.github.hdzitao.editstarters.ui.swing;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 通用table set/get value
 *
 * @version 3.2.1
 */
public class TableValue {
    private final int columnMax;
    private final Map<Integer, Setter> setterMap;
    private final Map<Integer, Getter> getterMap;

    /**
     * setter
     */
    public static interface Setter {
        void set(int row, int column, Object value);
    }

    /**
     * getter
     */
    public static interface Getter {
        Object get(int row, int column);
    }

    public TableValue(int columnMax) {
        this.columnMax = columnMax;

        this.setterMap = new ConcurrentHashMap<>(columnMax);
        this.getterMap = new ConcurrentHashMap<>(columnMax);
    }

    /**
     * 设置getter/setter
     */
    public void putValuer(int column, Getter getter) {
        putValuer(column, getter, null);
    }

    /**
     * 设置getter/setter
     */
    public void putValuer(int column, Getter getter, Setter setter) {
        if (column >= 0 && column < columnMax) {
            if (getter != null) {
                getterMap.put(column, getter);
            }

            if (setter != null) {
                setterMap.put(column, setter);
            }
        }
    }

    /**
     * 包含setter
     */
    public boolean hasSetter(int column) {
        return setterMap.containsKey(column);
    }

    /**
     * 获取值
     */
    public Object getValueAt(int row, int column) {
        Getter getter = getterMap.get(column);
        if (getter == null) {
            return null;
        }

        return getter.get(row, column);
    }

    /**
     * 设置值
     */
    public void setValueAt(Object value, int row, int column) {
        Setter setter = setterMap.get(column);
        if (setter == null) {
            return;
        }

        setter.set(row, column, value);
    }
}

File: src/main/kotlin/io/github/hdzitao/editstarters/ui/swing/TableValue.kt
package io.github.hdzitao.editstarters.ui.swing

import com.jetbrains.rd.util.concurrentMapOf


private typealias Getter = (Int, Int) -> Any
private typealias Setter = (Int, Int, Any) -> Unit

/**
 * 通用table set/get value
 *
 * @version 3.2.1
 */
class TableValue(private val columnMax: Int) {
    private val setterMap: MutableMap<Int, Setter> = concurrentMapOf()
    private val getterMap: MutableMap<Int, Getter> = concurrentMapOf()

    /**
     * 设置getter/setter
     */
    fun putValuer(column: Int, getter: Getter?, setter: Setter? = null) {
        if (column in 0..<columnMax) {
            if (getter != null) {
                getterMap[column] = getter
            }

            if (setter != null) {
                setterMap[column] = setter
            }
        }
    }

    /**
     * 包含setter
     */
    fun hasSetter(column: Int): Boolean {
        return setterMap.containsKey(column)
    }

    /**
     * 获取值
     */
    fun getValueAt(row: Int, column: Int): Any? {
        val getter = getterMap[column] ?: return null

        return getter(row, column)
    }

    /**
     * 设置值
     */
    fun setValueAt(value: Any, row: Int, column: Int) {
        val setter = setterMap[column] ?: return

        setter(row, column, value)
    }
}
